Fix mirror dimensions when constrainDimensions is active and not using fixed item width

dsandstrom opened the pull request:
When clicking an item, the mirror is added to the container.
Calculations are then made to figure out what size the mirror should be.
However, since the mirror is already in the container, the container now
has an extra item, thus reducing the width of all of the container's items.
Hence, the width calculated is too small and doesn't reflect the true
source size.

Instead, let's initially hide the mirror, figure out the intended width,
then show the mirror.

### This PR closes the following issues... _(if applicable)_

Maybe fixes #132, but not sure if the same issue

### Does this PR require the Docs to be updated?

don't think so

### Does this PR require new tests?

tests?

### This branch been tested on... _(click all that apply / add new items)_

**Browsers:**

* [x] Chrome latest

dsandstrom commented:
Let me know about tests.

tsov commented on src/Draggable/Draggable.js:
```diff
@@ -395,6 +395,7 @@ export default class Draggable {
     if (!isDragEvent(originalEvent) && !mirrorCreateEvent.canceled()) {
       const appendableContainer = this[getAppendableContainer]({source: this.originalSource});
       this.mirror = this.source.cloneNode(true);
+      this.mirror.style.display = 'none';
```
Could you move this line to `Mirror.js`, inside the `onMirrorCreated` method? That way the logic for showing hiding is in the same module. We are also currently working on moving all mirror related logic from `Draggable` to `Mirror`

dsandstrom commented:
Will do, thanks for the review.

dsandstrom commented:
@tsov I moved everything to Mirror.js as requested.  Also, I rebased and collapsed my commits.

I think I need some help with tests.  I tried adding something here: https://github.com/Shopify/draggable/blob/c4c98f830553e118fd3cc4a5c64f5f57edd3ff63/src/Draggable/Plugins/Mirror/tests/Mirror.test.js#L138 , but I must be doing something wrong.  The `mirrorElement` doesn't show the style attribute that I added, so I'm not sure how to test it.

I tried:
```js
expect(mirrorElement.style.display).toEqual('none');
```

MindSculpt commented:
@tsov @dsandstrom Can someone merge this change into master? If you use percentage-based divs (e.g. I'm using flexbox) as horizontal draggable items, the dragged item's width doesn't match the original (it's smaller). If there's another way to fix this other than this pull request please let me know.

zjffun commented:
@MindSculpt  There is a similar bug https://github.com/Shopify/draggable/pull/391 , and I try to fix this by using `position: fixed`

dsandstrom commented:
Looks like we need to ping @fdietze now.  Should I rebase this one?

MindSculpt commented:
@zjffun Yeah I saw that, and it definitely fixes the mirror size, but dragging doesn't seem to work 100% of the time in sortable. I'm using sortable for dragging percentage-based flexbox items horizontally, and even with your fix you occasionally have to click to drag an item more than once to get it to pick up.

fdietze commented:
> Looks like we need to ping @fdietze now. Should I rebase this one?

When you agree that this PR is ready, I guess you can merge it yourself. Many people have write access now.

zjffun commented:
@MindSculpt Thankes for test. I tried this fix, it can fix #391 also.

MindSculpt commented:
@zjffun Can we get this merged into the main repo? Your `position: fixed` fix doesn't seem to work consistently when you sort elements (swap positions) of different widths horizontally.

zjffun commented:
@MindSculpt I think this can merged into the main repo.
@fdietze Could you rebase and force push it?

fdietze commented:
> @fdietze Could you rebase and force push it? Please ignore my merge commit.

Don't you have the permissions to do it, or what is the problem?

zjffun commented:
@fdietze  I made a mistake, I want to at @dsandstrom. I have permissions to push on the branch of this pr and force push a rebase one, but it stuck in circleci now. I will check the reason later.

I am not very familiar with maintenance a collaboration projects. Is it appropriate to rebase someone else's branch? Or the rebase operations should be done by themselves?

dsandstrom commented:
@zjffun My branch is already based on master so looks like you rebased it properly.  Cirlceci looks to have passed, so not sure what the hold up is.  Maybe try re-running the job.

fdietze commented:
> I am not very familiar with maintenance a collaboration projects. Is it appropriate to rebase someone else's branch? Or the rebase operations should be done by themselves?

If you communicate with the person, that's fine. 

I recommend to use fore-with-lease when force pushing: https://stackoverflow.com/questions/52823692/git-push-force-with-lease-vs-force#:~:text=force%20overwrites%20a%20remote%20branch,elses%20work%20by%20force%20pushing.

zjffun commented:
According to [Workflow status checks never completes because of: ci/circleci — Waiting for status to be reported – CircleCI Support Center](https://support.circleci.com/hc/en-us/articles/360004346254-Workflow-status-checks-never-completes-because-of-ci-circleci-Waiting-for-status-to-be-reported) , we should unchecked `ci/circleci`, but I have no permissions to do this. 

I tested uncheck `ci/circleci` (`ci/circleci` option seems already removed from github branch protection rule now) in my repo it work fine, https://github.com/zjffun/draggable/pull/1 , https://github.com/zjffun/draggable/pull/2 .

Could you change the branch protection rule? @tsov @beefchimi

MindSculpt commented:
@tsov @beefchimi @zjffun @fdietze @dsandstrom Thanks everyone, this is working perfectly now for percentage-based width items in sortable.

tsov commented:
Also to keep in mind: This repo will automatically do NPM releases when the package verison gets bumped in master. Let me know if you want to do a version bump too!
